walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,122 @@
+import java.util.Collections;
+import java.util.List;
 
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    //TODO (@zous): Clean up comments.
+
+  // Runtime: O(n^2) where n is number of events passed in. 
+  // Ideally this problem could be dealt in O(n) time with the events sorted before 
+  // finding timeslots. 
+  public Collection<TimeRange> query (Collection<Event> events, MeetingRequest request){
+        // For optional attendees, just run query twice. First with optional  attendees included, if that is 
+        // not empty then return. If empty, then query just with mandatory attendees. 
+      Collection<String> allAttendees = new ArrayList<>();
+      allAttendees.addAll(request.getAttendees());
+      allAttendees.addAll(request.getOptionalAttendees());
+      MeetingRequest requestAll = new MeetingRequest (allAttendees, request.getDuration()); 
+      Collection<TimeRange> solutionsAll = queryHelper(events, requestAll);
+      // Return query of with optional guests, if there are options
+      // or if there are no mandatory attendees.
+      if (!(solutionsAll.isEmpty()) || request.getAttendees().isEmpty()){
+        return solutionsAll;
+      }
+      else{
+          return queryHelper(events, request);
+      }
+  }
+
+  //Runtime: O(n^2) where n is the size of the events. 
+  private Collection<TimeRange> queryHelper(Collection<Event> events, MeetingRequest request) {
+    // Want to return a collection of time ranges that work for the given request.
+    ArrayList<TimeRange> solutions = new ArrayList<>();
+    // No options for too long of a requests.
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()){
+      return solutions;
+    }
+    solutions.add(TimeRange.WHOLE_DAY);
+    // Process case where duration == 0. Would also return whole day. 
+    // If there are no attendees then return WHOLE_DAY.
+    if (request.getAttendees().isEmpty() || request.getDuration() == 0){
+      return solutions;
+    }
+    for (Event event: events){
+        // Check if event attendees are in the request. If not continue to next event.
+        boolean relevantEvent = overlappingAttendees(request, event);
+        if (!relevantEvent){
+            continue;
+        }
+        // Take out conflicting time from current answer of TimeRanges.
+        solutions = takeOutConflicts(solutions, event, request.getDuration());
+    }
+    return solutions;
+    //cannot conflict with the collection of events given.
+    //make sure to check end point of time range is inclusive. 
+  }
+
+  /**Splits a TimeRange into two time or shotens one TimeRange due to conflict with a given event */
+  // Runtime: O(n) where n is the size of the Old Solutions.
+  private ArrayList<TimeRange> takeOutConflicts(ArrayList<TimeRange> oldSolutions, Event event, long duration) {
```

Comment:
If |takeOutConflicts| is only being called once, I'm wondering why it's a separate function rather than being the bulk of the logic within |queryHelper|, which has the more ambiguous function name.